clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/service/servergroup/OracleBMCSServerGroupPersistence.groovy

```diff
@@ -0,0 +1,257 @@
+
+  /**
+   * Lists the server group names for the specified account.
+   *
+   * We do not consider "list" to conflict with READ, UPSERT or DELETE; this is because we only look at the names and
+   * to get the actual data a client has to subsequently perform READ operations anyway.
+   *
+   * The intended usage by the client of this class is:
+   *
+   * - list the server group names;
+   * - for each name, do a read and get the actual data;
+   * - if someone has deleted the server group sometimes between the list and the read, the client deals with it.
+   */
+  protected List<String> listServerGroupNames(OracleBMCSPersistenceContext ctx) {
+    def result = []
+    def namespace = getStorageNamespaceFor(ctx)
+    ensureStorageBucketFor(ctx)
+
+    try {
+      def rq = ListObjectsRequest.builder().namespaceName(namespace)
+        .bucketName(SERVERGROUP_BUCKET_NAME)
+        .build()
+      def rs = ctx.creds.objectStorageClient.listObjects(rq)
+      def maybeListSummaries = rs?.getListObjects()?.getObjects()
+      if (maybeListSummaries != null) {
+        result = maybeListSummaries.collect { it.getName() }
+      }
+    }
+    catch (Exception e) {
+      throw new OracleBMCSServerGroupPersistenceException("Failed listing contents of bucket", e)
+    }
+
+    return result
+  }
+
+  /**
+   * Reads the server group data with the provided name in the provided account.
+   */
+  OracleBMCSServerGroup getServerGroupByName(OracleBMCSPersistenceContext ctx, String name) {
+    return doPersistenceOperation(PersistenceOperation.READ, ctx, null, name)
+  }
+
+  /**
+   * Writes the server group data to the persistent store; the account is inferred from the server group data.
+   */
+  void upsertServerGroup(OracleBMCSServerGroup sg) {
+    if (sg.credentials.compartmentId == sg.launchConfig["compartmentId"] as String) {
+      doPersistenceOperation(PersistenceOperation.UPSERT, new OracleBMCSPersistenceContext(sg.credentials), sg, sg.name)
+    } else {
+      throw new OracleBMCSServerGroupPersistenceException("Different compartments - this is not allowed")
+    }
+  }
+
+  /**
+   * Deletes the server group data from the persistent store; the account is inferred from the server group data.
+   */
+  void deleteServerGroup(OracleBMCSServerGroup sg) {
+    if (sg.credentials.compartmentId == sg.launchConfig["compartmentId"] as String) {
+      doPersistenceOperation(PersistenceOperation.DELETE, new OracleBMCSPersistenceContext(sg.credentials), sg, sg.name)
+    } else {
+      throw new OracleBMCSServerGroupPersistenceException("Different compartments - this is not allowed")
+    }
+  }
+
+  @Synchronized
+  private String getStorageNamespaceFor(OracleBMCSPersistenceContext ctx) {
+    if (ctx.namespace) {
+      return ctx.namespace
+    }
+    def rq = GetNamespaceRequest.builder().build()
+    def rs = ctx.creds.objectStorageClient.getNamespace(rq)
+    if (rs?.getValue() == null) {
```

Comment:
Just a nit, but could capture result of `rs?.getValue()` so we call it just the one time instead of 3 by the end of the method.